File: samplesJS/TiledMap/assets/scripts/TiledMapSample.js
var tiledMap = getTiledMap("sample.tmx");
var ZOOMED_TRANSFORM = Matrix.create2DTranslationZoom(new Vector3(512, 384, 0), 2);
var tileSize = tiledMap.getTileSize();
var halfTileSize = tileSize * .5;
var shader = getShader("sample.ps");

var path1 = tiledMap.getPath(13, 27, 6, 12, PathType.CROSS_CORNERS);
var path2 = tiledMap.getPath(13, 27, 6, 12, PathType.STRAIGHT);
var path3 = tiledMap.getPath(13, 27, 6, 12, PathType.ALLOW_DIAGONAL);

var OPTION_NORMAL = 0;
var OPTION_ZOOMED = 1;
var OPTION_SHADER = 2;
var option = OPTION_NORMAL;

function drawPath(path, color, transform)
{
    PrimitiveBatch.begin(PrimitiveMode.LINE_STRIP, null, transform);
    for (var i = 0; i < path.length; ++i)
    {
        var pos = path[i];
        PrimitiveBatch.draw(pos.mul(tileSize).add(halfTileSize), color);
    }
    PrimitiveBatch.end();
}

function update(dt)
{

}

function render() 
{
    var transform = Matrix.IDENTITY;

    // Render full map
    if (option == OPTION_ZOOMED)
    {
        transform = ZOOMED_TRANSFORM;

        // By wraping a sprite batch around it, we can apply transform
        SpriteBatch.begin(transform);
        tiledMap.render();
        SpriteBatch.end();
    }
    else if (option == OPTION_SHADER)
    {
        // By wraping a sprite batch around it, we can apply shaders
        SpriteBatch.begin(transform, shader);
        tiledMap.renderLayer(0);
        SpriteBatch.end();

        tiledMap.renderLayer(1);
        tiledMap.renderLayer(2);
        tiledMap.renderLayer(3);
    }
    else
    {
        // Simple map render, no need to activate the sprite batch
        tiledMap.render();
    }

    // Draw paths
    drawPath(path1, new Color(1, 0, 0), transform);
    drawPath(path2, new Color(0, 1, 0), transform);
    drawPath(path3, new Color(0, 0, 1), transform);
}

function renderUI()
{
    if (GUI.begin("Options"))
    {
        option = GUI.radioButton("Normal", option, 0);
        option = GUI.radioButton("Zoomed", option, 1);
        option = GUI.radioButton("Shaders to bottom layer", option, 2);
    }
    GUI.end();
}

File: samples/TiledMap/src/TiledMapSample.cpp
// Oak Nut include
#include <onut/PrimitiveBatch.h>
#include <onut/Renderer.h>
#include <onut/Settings.h>
#include <onut/TiledMap.h>

OTiledMapRef pTiledMap;
OTiledMap::Path path1;
OTiledMap::Path path2;
OTiledMap::Path path3;
OTiledMap::Path path4;

void initSettings()
{
    oSettings->setGameName("Tiled Map Sample");
    oSettings->setResolution(Point(1024, 768));
}

void init()
{
    pTiledMap = OGetTiledMap("sample.tmx");
    pTiledMap->generateCollisions("collisions");
    path1 = pTiledMap->getPath(Point(13, 27), Point(6, 12), OTiledMap::PATH_ALLOW_DIAGONAL | OTiledMap::PATH_CROSS_CORNERS);
    path2 = pTiledMap->getPath(Point(13, 27), Point(6, 12), 0);
    path3 = pTiledMap->getPath(Point(13, 27), Point(6, 12), OTiledMap::PATH_ALLOW_DIAGONAL);
    path4 = pTiledMap->getPath(Point(13, 27), Point(1, 30), OTiledMap::PATH_ALLOW_DIAGONAL | OTiledMap::PATH_CROSS_CORNERS); // Impossible, will get closest
}

void update()
{
}

void drawPath(const OTiledMap::Path& path, const Matrix& transform, const Color& color)
{
    auto tileSize = (float)pTiledMap->getTileSize();
    auto halfTileSize = tileSize * .5f;

    oPrimitiveBatch->begin(OPrimitiveLineStrip, nullptr, transform);
    for (const auto& p : path)
    {
        oPrimitiveBatch->draw(Vector2((float)p.x * tileSize + halfTileSize, (float)p.y * tileSize + halfTileSize), color);
    }
    oPrimitiveBatch->end();
}

void render()
{
    auto transform = Matrix::Create2DTranslationZoom(OScreenf, { 512, 384 });

    pTiledMap->setTransform(transform);
    pTiledMap->render();

    drawPath(path1, transform, Color(1, 0, 0));
    drawPath(path2, transform, Color(0, 1, 0));
    drawPath(path3, transform, Color(0, 0, 1));
    drawPath(path4, transform, Color(1, 0, 1));
}

void postRender()
{
}

void renderUI()
{
}
